aws-api/src/main/java/com/amplifyframework/api/aws/MutiAuthSubscriptionOperation.java

```diff
@@ -0,0 +1,255 @@
+    }
+
+    private void dispatchRequest() {
+        LOG.debug("Processing subscription request: " + getRequest().getContent());
+        if (authTypes.hasNext()) {
+            AuthorizationType authorizationType = authTypes.next();
+            LOG.debug("Attempting to subscribe with " + authorizationType.name());
+            GraphQLRequest<T> request = getRequest();
+            // if the rule we're currently processing is an owner-based rule,
+            // then call the AuthRuleRequestDecorator to see if the owner needs to be
+            // added to the request.
+            if (authTypes.isOwnerBasedRule()) {
+                try {
+                    request = requestDecorator.decorate(request, authorizationType);
+                } catch (ApiException apiException) {
+                    LOG.warn("Unable to automatically add an owner to the request.", apiException);
+                    if (apiException instanceof ApiAuthException) {
+                        subscriptionFuture = executorService.submit(this::dispatchRequest);
+                    } else {
+                        emitErrorAndCancelSubscription(apiException);
+                    }
+                    return;
+                }
+            }
+            subscriptionEndpoint.requestSubscription(
+                request,
+                authorizationType,
+                subscriptionId -> {
+                    MutiAuthSubscriptionOperation.this.subscriptionId = subscriptionId;
+                    onSubscriptionStart.accept(subscriptionId);
+                },
+                item -> {
+                    if (item.hasErrors()) {
+                        for (GraphQLResponse.Error error : item.getErrors()) {
+                            if (error.getExtensions() != null &&
+                                UNAUTHORIZED_EXCEPTION.equals(error.getExtensions().get("errorType"))) {
+                                subscriptionFuture = executorService.submit(this::dispatchRequest);
+                                return;
+                            }
+                        }
+                        emitErrorAndCancelSubscription(new ApiException("The server returned subscription errors:" +
+                                                                        item.getErrors().toString(),
+                                                                    AmplifyException.TODO_RECOVERY_SUGGESTION));
+                    } else {
+                        onNextItem.accept(item);
+                    }
+                },
+                apiException -> {
+                    LOG.warn("A subscription error occurred.", apiException);
+                    if (apiException instanceof ApiAuthException) {
+                        executorService.submit(this::dispatchRequest);
+                    } else {
+                        emitErrorAndCancelSubscription(apiException);
+                    }
```

Comment:
I'm slightly confused about the `UNAUTHORIZED_EXCEPTION` vs `ApiAuthException`.

So `ApiAuthException` is when subscription connection fails to be established in the first place due to permission, and `UNAUTHORIZED_EXCEPTION` is from a partial error in GraphQL response right? How would multi-auth solve both issues? isn't it only capable of resolving auth-related errors in GraphQL responses (the latter)?

Reply:
`UNAUTHORIZED_EXCEPTION` is meant to be used when we are inspecting the body of a response received on the websocket. This is needed because not all errors are emitted as a `subscription_error` by AppSync. So when we get a successful response, we need to inspect the body and check for that.

That being said, rather than doing this at the operation level, I moved the logic of checking for that into the `SubscriptionEndpoint`. If a message with an `UNAUTHORIZED_EXCEPTION` is received, we're now emitting an `ApiAuthException`.

The purpose behind `ApiAuthException` is to differentiate from other API-related exceptions. It makes it easier to determine if we need to queue up a retry or if the error received was not auth-related and we should just bail.